TST: Azure CI for Windows testing

Since our Appveyor runs are rather slow I've been working on getting Azure CI working for SciPy Windows testing.

To reduce noise / CI burden I've been doing this in a [PR in my fork](https://github.com/tylerjereddy/scipy/pull/20), where you can see the current diff & Azure CI results.

I've struggled for several hours trying to get openblas to work properly--I note that [our instructions for installing openblas](https://scipy.github.io/devdocs/building/windows.html#installing-openblas) point to a `FOUND` output with `language = c`, while `language = f77` in current appveyor builds, but I see `c` in my current build attempts on Azure (and the library name gets duplicated in the list on Azure too).

There are also apparent differences in the files / static libraries used for our master branch appveyor runs vs. macpython wheel building on appveyor & I've tried copying the static libraries from both URLs without full success yet.

The other question is whether the pre-installed mingw / gfortran on the Azure images is compatible with this "just copy the static library(ies)" from openblas workflow.

The Appveyor file suggests that we should `install OpenBLAS.dll to MINGW\bin`, but actually doesn't appear to do this in our master branch CI.

I'm also not entirely clear on why we need to replace NumPy distutils in our builds using the master branch--is this still true / helpful? Presumably the changes to distutils will eventually show up in a release and we can remove that stuff?

A persistent problem that I can reproduce locally for Windows source builds after openblas is detected by distutils is this mingw gfortran (linking?) issue:

`C:\Users\treddy\windows_dev_work\github_projects\scipy/scipy/integrate/quadpack/dqc25f.f:293: undefined reference to 'dgtsv_'`

[There's a line](https://github.com/scipy/scipy/blob/master/appveyor.yml#L104) in our master branch appveyor config that suggest that the opeblas dll may simply be copied into the mingw gfortran `bin` path, but this seems to have no effect for me (same for copying various lib/ and include/ files over to corresponding mingw paths). I've even tried messing around with the gcc and gfortran `specs` for lib path searching.

--- a/azure-pipelines.yml
+++ b/azure-pipelines.yml
@@ -1,0 +1,110 @@
+trigger:
+  # start a new build for every push
+  batch: False
+  branches:
+    include:
+      - master
+      - maintenance/*
+
+jobs:
+- job: Windows
+  pool:
+    vmIMage: 'VS2017-Win2016'
+  variables:
+      OPENBLAS_32: https://3f23b170c54c2533c070-1c8a9b3114517dc5fe17b7c3f8c63a43.ssl.cf2.rackcdn.com/openblas-5f998ef_gcc7_1_0_win32.zip
+      OPENBLAS_64: https://3f23b170c54c2533c070-1c8a9b3114517dc5fe17b7c3f8c63a43.ssl.cf2.rackcdn.com/openblas-5f998ef_gcc7_1_0_win64.zip
+  strategy:
+    maxParallel: 4
+    matrix:
+        Python36-32bit-full:
+          PYTHON_VERSION: '3.6'
+          PYTHON_ARCH: 'x86'
+          TEST_MODE: full
+          OPENBLAS: $(OPENBLAS_32)
+          BITS: 32
+        Python35-64bit-full:
+          PYTHON_VERSION: '3.5'
+          PYTHON_ARCH: 'x64'
+          TEST_MODE: full
+          OPENBLAS: $(OPENBLAS_64)
+          BITS: 64
+        Python36-64bit-full:
+          PYTHON_VERSION: '3.6'
+          PYTHON_ARCH: 'x64'
+          TEST_MODE: full
+          OPENBLAS: $(OPENBLAS_64)
+          BITS: 64
+        Python37-64bit-full:
+          PYTHON_VERSION: '3.7'
+          PYTHON_ARCH: 'x64'
+          TEST_MODE: full
+          OPENBLAS: $(OPENBLAS_64)
+          BITS: 64
+  steps:
+  - task: UsePythonVersion@0
+    inputs:
+      versionSpec: $(PYTHON_VERSION)
+      addToPath: true
+      architecture: $(PYTHON_ARCH)
+  # as noted by numba project, currently need
+  # specific VC install for Python 2.7
+  - powershell: |
+      $wc = New-Object net.webclient
+      $wc.Downloadfile("https://download.microsoft.com/download/7/9/6/796EF2E4-801B-4FC4-AB28-B59FBF6D907B/VCForPython27.msi", "VCForPython27.msi")
+      Start-Process "VCForPython27.msi" /qn -Wait
+    displayName: 'Install VC 9.0'
+    condition: eq(variables['PYTHON_VERSION'], '2.7')
+  - script: python -m pip install --upgrade pip setuptools wheel
+    displayName: 'Install tools'
+  - powershell: |
+      $wc = New-Object net.webclient;
+      $wc.Downloadfile("$(OPENBLAS)", "openblas.zip")
+      $tmpdir = New-TemporaryFile | %{ rm $_; mkdir $_ }
+      Expand-Archive "openblas.zip" $tmpdir
+      $pyversion = python -c "from __future__ import print_function; import sys; print(sys.version.split()[0])"
+      Write-Host "Python Version: $pyversion"
+      $target = "C:\\hostedtoolcache\\windows\\Python\\$pyversion\\$(PYTHON_ARCH)\\lib\\openblas.a"
+      Write-Host "target path: $target"
+      cp $tmpdir\$(BITS)\lib\libopenblas_5f998ef_gcc7_1_0.a $target
+    displayName: 'Download / Install OpenBLAS'
+  - powershell: |
+      # NOTE: can probably (eventually) abstract this 
+      # upstream in Microsoft repo to support x86 natively
+      choco install -y mingw --forcex86 --force
+    displayName: 'Install 32-bit mingw for 32-bit builds'
+    condition: eq(variables['BITS'], 32)
+  - script: python -m pip install numpy cython==0.28.5 pytest pytest-timeout pytest-xdist pytest-env pytest-faulthandler Pillow mpmath matplotlib
+    displayName: 'Install dependencies'
+  - powershell: |
+      # need a version of NumPy distutils that can build
+      # with msvc + mingw-gfortran
+      $NumpyDir = $((python -c 'import os; import numpy; print(os.path.dirname(numpy.__file__))') | Out-String).Trim()
+      rm -r -Force "$NumpyDir\distutils"
+      $tmpdir = New-TemporaryFile | %{ rm $_; mkdir $_ }
+      git clone -q --depth=1 -b master https://github.com/numpy/numpy.git $tmpdir
+      mv $tmpdir\numpy\distutils $NumpyDir
+    displayName: 'Replace NumPy distutils'
+  - powershell: |
+      If ($(BITS) -eq 32) {
+          # 32-bit build requires careful adjustments
+          # until Microsoft has a switch we can use
+          # directly for i686 mingw
+          $env:NPY_DISTUTILS_APPEND_FLAGS = 1
+          $env:CFLAGS = "-m32"
+          $env:LDFLAGS = "-m32"
+          $env:PATH = "C:\\tools\\mingw32\\bin;" + $env:PATH
+          refreshenv
+      }
+
+      mkdir dist
+      pip wheel --no-build-isolation -v -v -v --wheel-dir=dist .
+      ls dist -r | Foreach-Object {
+          pip install $_.FullName
+      }
+    displayName: 'Build SciPy'
+  - script: python runtests.py -n --mode=$(TEST_MODE) -- -n auto -rsx --junitxml=junit/test-results.xml
+    displayName: 'Run SciPy Test Suite'
+  - task: PublishTestResults@2
+    inputs:
+      testResultsFiles: '**/test-*.xml'
+      testRunTitle: 'Publish test results for Python $(python.version)'

